### ISSUE TITLE
Contradiction in computing derivatives of pdf in glossy BSDF

### ISSUE THREAD

**Issue by saeedhd96** [mitsuba-renderer/mitsuba2#539]
- [🐛 bug report]

## Summary

The pdf derivatives wrt roughness parameter are not equal when using bsdf.pdf() and bsdf.sample(). This happens on a glossy bsdf (e.g rough-conductor).

## System configuration

- Mitsuba 2 version: the current version of this repo
- Compiled variants:
    * `gpu_autodiff_rgb`

## Description

My problem is with using bsdf.pdf() and bsdf.sample() in order to compute the derivative of pdf wrt roughness parameter in a glossy bsdf (e.g rough-conductor). Not sure if this is a bug or my understanding of these functions is not correct.

My experiment was that for a rough conductor copper,  using a given w_i, I used bsdf.sample() to sample "w_o, bsdf_val and pdf". Then, using the the sampled "w_o", I used bsdf.pdf() to get "pdf". The resulting "pdf" is exactly equal to the "pdf" from bsdf.sample(). However, their gradients wrt roughness parameter is not equal.

## Steps to reproduce
1. In a cornell box when the red wall is copper rough conductor with alpha 0.3, we set up roughness parameter as differentiable:

```
param_0 = params['copper.alpha.value']
params.keep(['copper.alpha.value'])
ek.set_requires_gradient(param_0)

```

2. Gradients of pdf from bsdf.sample():

```
bs, bsdf_val = BSDF.sample_vec(bsdf, ctx, si, sampler.next_1d(active), sampler.next_2d(active), active)
ek.set_gradient(param_0,1,backward= False)
Float.forward()
print(bs.pdf)
print(ek.gradient(bs.pdf))
```

Output:

```
[1.42621, 0.918134, 0.275155, 0.388981, 1.1572, .. 310 skipped .., 1.54476, 0.834284, 0.28909, 0.340116, 0.0775129]
[-10.085, -5.98488, -1.14211, -2.1575, -7.62232, .. 310 skipped .., -10.1419, -5.27179, -1.80803, -1.61196, -0.363747]

```
3. Gradients of bsdf.pdf() using bs.wo from the previous step:

```
pdf = BSDF.pdf_vec(bsdf, ctx, si, bs.wo , active)
ek.set_gradient(param_0,1,backward= False)
Float.forward()
print(pdf)
print(ek.gradient(pdf))

```

```
[1.42621, 0.918133, 0.275155, 0.388981, 1.1572, .. 310 skipped .., 1.54476, 0.834284, 0.28909, 0.340116, 0.077513]
[3.53934, 0.256007, 1.28402, 1.54916, -7.1008, .. 310 skipped .., -10.1921, -0.181845, 5.10267, 0.649189, 1.77677]
```

As you can see, the pdf values are the same, but their gradients are different.

**Comment by wjakob** [comment]
Hi Saeed -- did you try this on mitsuba2-next (which IIRC you have access to.)
Many AD-related issues have been fixed in the new codebase, so this issue might be gone there.

**Comment by saeedhd96** [comment]
Hi Wenzel,

I wanted to, but I guess my "watch" access was not enough to pull the repo. I would appreciate the required access to donwload mitsuba-next. Also, not sure if this is a bug or is an inequality due to detached/attached Monte Carlo samples.

**Comment by saeedhdd** [comment]
I tried this on mitsuba-next and it is fortunately fixed; we can close this issue. Thanks @wjakob!
